uid/snowid.go

```diff
@@ -0,0 +1,192 @@
+// node
+type Node struct {
+	mu    sync.Mutex
+	epoch time.Time
+	time  int64
+	node  int64
+	step  int64
+
+	nodeMax   int64
+	nodeMask  int64
+	stepMask  int64
+	timeShift uint8
+	nodeShift uint8
+}
+
+// An ID is a custom type used for a snowflake ID.  This is used so we can
+// attach methods onto the ID.
+type ID int64
+
+// NewNode returns a new snowflake node that can be used to generate snowflake
+// IDs
+func NewNode(node int64) (*Node, error) {
+	n := Node{}
+	n.node = node
+	n.nodeMax = -1 ^ (-1 << nodeBits)
+	n.nodeMask = n.nodeMax << stepBits
+	n.stepMask = -1 ^ (-1 << stepBits)
+	n.timeShift = nodeBits + stepBits
+	n.nodeShift = stepBits
+
+	if n.node < 0 || n.node > n.nodeMax {
+		return nil, errors.New("Node number must be between 0 and " + strconv.FormatInt(n.nodeMax, 10))
+	}
+
+	var curTime = time.Now()
+	// add time.Duration to curTime to make sure we use the monotonic clock if available
+	n.epoch = curTime.Add(time.Unix(epoch/1000, (epoch%1000)*1000000).Sub(curTime))
+
+	return &n, nil
+}
+
+// Generate creates and returns a unique snowflake ID
+// To help guarantee uniqueness
+// - Make sure your system is keeping accurate system time
+// - Make sure you never have multiple nodes running with the same node ID
+func (n *Node) Generate() ID {
+	n.mu.Lock()
+
+	now := time.Since(n.epoch).Nanoseconds() / 1000000
+
+	if now == n.time {
+		n.step = (n.step + 1) & n.stepMask
+
+		if n.step == 0 {
+			for now <= n.time {
+				now = time.Since(n.epoch).Nanoseconds() / 1000000
+			}
+		}
+	} else {
+		n.step = 0
+	}
+
+	n.time = now
+
+	r := ID((now)<<n.timeShift |
+		(n.node << n.nodeShift) |
+		(n.step),
+	)
+
+	n.mu.Unlock()
+	return r
+}
+
+// Bytes returns a byte array of the base58 encoded value for this ID.
+func (id ID) Bytes() []byte {
+	switch {
+	case id <= 0:
+		return nil
+	case id < 58:
+		return []byte{encodeBase58Map[id]}
+	}
+
+	b := make([]byte, 0, 11)
+	for id >= 58 {
+		b = append(b, encodeBase58Map[id%58])
+		id /= 58
+	}
+	b = append(b, encodeBase58Map[id])
+
+	for x, y := 0, len(b)-1; x < y; x, y = x+1, y-1 {
+		b[x], b[y] = b[y], b[x]
+	}
+
+	return b
+}
+
+// String returns the base58 encoded value for this ID.
+func (id ID) String() string {
+	return string(id.Bytes())
+}
+
+// Parse parses a base58 encoded value into an ID.
+func Parse(b []byte) (ID, error) {
+	switch {
+	case bytes.HasPrefix(b, []byte("1")):
+		return -1, fmt.Errorf("invalid base58: not in canonical form")
+	case len(b) > 11:
+		return -1, fmt.Errorf("invalid base58: too long")
+	}
+
+	var id int64
+	for i := range b {
+		if decodeBase58Map[b[i]] == 0xFF {
+			return -1, fmt.Errorf("invalid base58: byte %d is out of range", i)
+		}
+
+		shifted, ok := multiplyCheckOverflow(id, 58)
+		if !ok {
+			return -1, fmt.Errorf("invalid base58: value too large")
+		}
+		id = shifted + int64(decodeBase58Map[b[i]])
+		if id <= 0 {
+			return -1, fmt.Errorf("invalid base58: value too large")
+		}
```

Comment:
These two overflow checks are new. It took me a while to figure out the multiply overflow because it often doesn't result in a negative number. Eventually I found enough examples using the fuzz test to understand what was happening.

Reply:
The generation should not have this problem. It should never generate an `int64` that is out of range. We don't have a proper fuzz test for that yet. I think the `if < 0` check in `TestNode_Generate_ConcurrencyAndUniqueness` should confirm that generation always generates valid IDs that don't overflow. We could add a fuzz test for it as well.